core/src/main/java/org/elasticsearch/action/DocWriteResponse.java

```diff
@@ -31,30 +31,78 @@
 import org.elasticsearch.rest.RestStatus;
 
 import java.io.IOException;
+import java.util.Locale;
 
 /**
  * A base class for the response of a write operation that involves a single doc
  */
 public abstract class DocWriteResponse extends ReplicationResponse implements WriteResponse, StatusToXContent {
 
+    public enum Operation {
+        CREATE(0),
+        INDEX(1),
+        DELETE(2),
+        NOOP(3);
+
+        private final byte op;
+        private final String lowercase;
+
+        Operation(int op) {
+            this.op = (byte) op;
+            this.lowercase = this.toString().toLowerCase(Locale.ENGLISH);
+        }
+
+        public byte getOp() {
+            return op;
+        }
+
+        public String getLowercase(){
+            return lowercase;
+        }
+
+        public static final Operation formOperation(byte opcode) {
```

Comment:
I think you mean `fromOperation`?

I think it'd be slightly better to have `Operation` implement `Writeable` and call this method `readFrom` and have it take the `StreamInput` instead. At least, that is how we've been doing these enums in the past few months.